Features/task 1

milan009 opened the pull request.

Commit:
KA-182 Create VS solution

Commit:
KA-183 Add ListItem model

Commit:
KA-183 Install Microsoft.AspNet.WebApi.Versioning NuGet package

Commit:
KA-183 Add Items controller, implement URL versioning

Commit:
KA-183 Implement the HTTP methods, revision versioning

Commit:
KA-183 Refactor Put method

When item with given ID is not found, a new resource is created

Commit:
KA-183 Added NUnit package

Commit:
KA-183 Change string IDs to instances of Guid

Commit:
KA-183 Update VS compilers to allow using C#7 features

Commit:
KA-183 Reformat version routing, bind POST param from request

Commit:
KA-183 Implement a way to inject IDGenerator method via constructor

Commit:
KA-183 Make PUT replace entire entity to behave as intended

Commit:
KA-183 Implement DELETE verb

Commit:
KA-183 Install JsonPatch package, implement PATCH verb

Commit:
KA-183 Fix PATCH implementation

Correct formatter had to be added. It is using case insensitive resolver now to allow better flexibility.

Commit:
KA-183 Refactored controller methods to be async

As the dummy controller does not perform any real asynchronous work, the methods just await a task that behaves exactly like the former synchronous methods.

Commit:
KA-183 Fix PUT method

I was not storing the newly created item into the list

Commit:
KA-183 Remove unused "using" statements

Commit:
KA-183 Add NUnit package, add empty unit test

Commit:
KA-183 Add GET method unit test, implement IEquatable

Quite big refactoring to the model had to be done in order
to make the controller testable. The Equals method had to be overriden and because of that, the GetHashCode as well.

However, to make GetHashCode work as intended, Id needed to be readonly, which proved to be quite difficult to achieve. The XML formatter requires a propperty to have public setter method.

Eventually, I settled up for implementing an empty public setter and private readonly "_id" field. See comments for further explanation.

With regards to this change, a minor changes to the list items initializations had to be made in the tests and in the controller itself.

Commit:
KA-183 Implement ListItem comparer, simplify the model

Commit:
KA-183 Add unit tests for GET and POST, extract constants

Commit:
KA-183 Use the constants, check for null argument in POST

Commit:
KA-183 Remove unused imports

Commit:
KA-183 Implement PUT method for collection

Commit:
KA-183 Refactor GUID/url parsing and checking

Seems like the binder is capable of parsing to GUID directly, although I cannot controll the error response this way.

Commit:
KA-183 Make ID param "optional"

The ID param is always parsed from the URL - it will never be null except when the passed id param is not convertible to a GUID.

This approach gives a little more reasonable error message than the former as the error is stored in the ModelState.

Commit:
Revert "KA-183 Make ID param "optional""

This reverts commit f4a25bc108a515c0046c68a6942d6013fcc8f619.

Commit:
Revert "KA-183 Refactor GUID/url parsing and checking"

This reverts commit 41cca975a726462e83fe3895127263e3c736d062.

Commit:
KA-183 Add custom action filters and Guid binder

The goal of this commit was to remove validation logic from the controller methods.

The argument validation has been moved out from the controller into custom action filters. The three filters now take care of:
- Null arguments
- Model validity
- Consistency of GUIDs in Put request

All validation has been removed from the controller, the filters have been added in WebApiConfig (with exception of the GUID consistency filter, as it only matters to PUT request)

A custom GUID model binder has been created, as the default one was not correctly returning an error on invalid guid formats (a "null" value was passed to the ModelState dictionary - this did consequently cause an error, as the key of this value was not nullable; however, the resulting error message was not informative enough)

This model binder was then applied globally to all Guid type parameters in the WebApiConfig. To reflect on this change, all types of Id arguments were changed from String to Guid, and all checks whether the argument can be parsed were removed from the controller - the binder takes care of this functionality now.

All unit tests that were using string guid arguments will not work in this commit and need to be refactored to reflect the change!

Some comments were added to clarify code, some unused "using" statements were removed.

Commit:
KA-183 Refactor async methods simulation

Commit:
KA-183 Remove custom GuidModelBinder

The default Guid conversion handling seems good enough

Commit:
KA-183 Implement tests for GET invalid request, extract config init

The server/client initialization is required only for a small subset of tests - no reason to go through it in startup with every single unit test.

Commit:
KA-183 Change MockListItems to give deep copy of the list

The way I was using this "constant" required me to create deep copies of it in every situation, which I had not been doing; that lead to quite unexpected behavior.

Now, a completely new list is always created when demanded, bypassing the need alltogether.

Commit:
KA-183 Implement additional test, refactor code

As the controller uses a static list to store ListItems, the consequent execution of unit test led to unexpected behavior - the state from former test runs had been perserved.

A new public static method InitializeItems(), which resets the List to default mock value, has been created to compensate. Additionaly, a static constructor that uses this method to poppulate the List has been created.

Tests that use HTTP server/client have been refactored to correctly dispose the two with the "using" blocks.

Few oversights had been fixed:
- tests using the client/server approach are returning HttpResponseMessage instead of IHtttpActionResult; those tests have been changed accordingly.
- as the previous commit changed the way MockListItems was implemented and it no longer contains a '[]' accessor, 'ElementAt()' had to be used instead
- the "Post_Returns400OnNullText" had to be rewritten to use the client/server approach, as the null check is done by an action filter, not the controller directly

Commit:
KA-183 Add filter for collection checking

No longer is it possible to PUT empty collection. Also, the model attributes were flagged as "Required", so no more empty ListItems being PUT either.

Commit:
KA-183 Implement collection DELETE

Commit:
KA-183 Add tests for collection delete

Commit:
KA-183 Add PATCH filter, implement PATCH tests

Commit:
KA-183 Merge from experimental branch

tobiaskamenicky commented on ListApp/ListApp.Api.Tests/ItemsController.cs:
```diff
@@ -0,0 +1,349 @@
+    [TestFixture]
+    public class ItemsControllerTests
+    {
+        private static readonly Guid TheGuid = Guid.Parse("00000000-0000-0000-0000-000000000003");
```
you can generate random `Guid` so you don't have to write it...in VS->Tools->Create Guid-> select 4. option and copy

tobiaskamenicky commented on ListApp/ListApp.Api.Tests/ItemsController.cs:
```diff
@@ -0,0 +1,349 @@
+        public void SetUp()
+        {
+            ItemsController.InitializeItems();
+            _theController = new ItemsController(GuidCreator);
```
The name should be either _controller or _itemsController, don't use articles in naming.

tobiaskamenicky commented on ListApp/ListApp.Api/Utils/Constants.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        public static IEnumerable<ListItem> MockListItems => new List<ListItem>
+        {
+            new ListItem {Id = Guid.Parse("00000000-0000-0000-0000-000000000000"), Text = "Stretch correctly"},
```
It would be enough to store this directly in a controller since you will remove it in the next task.

tobiaskamenicky commented on ListApp/ListApp.Api/Utils/ListItemEqualityComparer.cs:
```diff
@@ -0,0 +1,28 @@
+    /// It does NOT implement its own GetHashCode() method - use ONLY
+    /// for comparing two <see cref="ListItem"/> models!
+    /// </summary>
+    public class ListItemEqualityComparer : IEqualityComparer<ListItem>
```
Since it is used only in tests, it shouldn't be part of the Api project.

tobiaskamenicky commented on ListApp/ListApp.Api.Tests/ItemsController.cs:
```diff
@@ -0,0 +1,349 @@
+        // The server/client has to be used to test this.
+        // The argument validity check is handled by the action filter
+        [Test]
+        public async Task Get_Returns400OnInvalidGuidFormat()
```
You can use names in format NameOfMethod_Parameters_WhatShouldHappen...in this case Get_InvalidGuid_ReturnsNotFound(). Using this naming, you can omit comments.

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+        {
+            // Currently used to simulate a way to store ListItems
+            private static List<ListItem> _items;
+            private readonly Func<Guid> _idGenerator;
```
This method should be moved to a separate service in the next task and the service should be injected.

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+
+            static ItemsController()
+            {
+                InitializeItems();
```
You can do this in a normal constructor. This method can be therefore private and your tests don't need to call it every time

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+
+            [Route]
+            [HttpGet]
+            public async Task<IEnumerable<ListItem>> GetItems()
```
Async methods should have Async at the end of the name, e.g. `GetItemsAsync()`.
The return type should be `Task<IHttpActionResult>`

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+            [HttpGet]
+            public async Task<IEnumerable<ListItem>> GetItems()
+            {
+                return await Task.FromResult<IEnumerable<ListItem>>(_items);
```
You don't have write type parameter explicitly...it is automatically resolved. The REST API should return status codes with objects...in this case `Ok(_items)`

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+
+            [Route]
+            [HttpPost]
+            public async Task<IHttpActionResult> PostItem([FromBody]string newItemText)
```
The parameter should be the entire model since you can add more properties in the future

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+
+                _items.Add(createdItem);
+                    
+                return await Task.FromResult<IHttpActionResult>(Created($"/items/{createdItem.Id}", createdItem));
```
The location shouldn't be created manually in the next task.

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+            [Route]
+            [HttpPut]
+            [PutCollectionActionFilter]
+            public async Task<IHttpActionResult> PutItemsCollection([FromBody] IEnumerable<ListItem> items)
```
By implementing batch actions, you not only complicate your work (with respect to future tasks) but it not very usual for REST APIs to have such actions. I would recommend removing all batch actions.

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+                {
+                    _items.Clear();
+                    _items.AddRange(listItems);
+                    return await Task.FromResult<IHttpActionResult>(Ok(_items));
```
I don't think this is a correct behavior of the PUT. PUT should either update existing or create new.

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+                }
+
+                return await Task.FromResult<IHttpActionResult>(Content(HttpStatusCode.NotFound,
+                    "One or more of IDs specfied for deletion has not been found."));
```
This message not necessary. It is implicated by the Status code ;)

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+                if (toDelete.All(idToDelete => _items.Exists(item => item.Id == idToDelete)))
+                {
+                    _items.RemoveAll(item => toDelete.Contains(item.Id));
+                    return await Task.FromResult<IHttpActionResult>(Ok());
```
Successful DELETE usually returns NoContent(), but this also fine

tobiaskamenicky commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,167 @@
+            [Route("{id}")]
+            [HttpPatch]
+            [PatchSingleResourceActionFilter]
+            public async Task<IHttpActionResult> PatchItem(Guid id, [FromBody] JsonPatch.JsonPatchDocument<ListItem> patch)
```
Not necessary method

tobiaskamenicky commented on ListApp/ListApp.Api/Filters/ModelValidationActionFilter.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace ListApp.Api.Filters
+{
+    public class ModelValidationActionFilter : ActionFilterAttribute
```
Great idea, even though it was not required!

tobiaskamenicky commented on ListApp/ListApp.Api/Filters/NullArgumentActionFilter.cs:
```diff
@@ -0,0 +1,19 @@
+            if (actionContext.ActionArguments.ContainsValue(null))
+            {
+                actionContext.Response =
+                    actionContext.Request.CreateErrorResponse(HttpStatusCode.BadRequest, "A null argument is not allowed!");
```
`actionContext.ModelState.AddErrror()` and return ModelState in a BadRequest? It could contain more errors then a single error value.

tobiaskamenicky commented on ListApp/ListApp.Api/Filters/ModelValidationActionFilter.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using System.Collections.Generic;
```
Unnecessary using directives...always check that or let R# solve it...Alt+Enter "Remove in Solution"

tobiaskamenicky commented on ListApp/ListApp.Api/Utils/ServerInit.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public static class ServerInit
+    {
+        public static HttpConfiguration InitializeConfiguration(HttpConfiguration config)
```
This should be written in style of `WebApiConfig.cs` and located in App_Start folder. = Register method that would be used in Global.asax

tobiaskamenicky replied:
Oh...you are using in tests...Is it really necessary? Could you solve the duplicate code that this contains (with WebApiConfig)?

tobiaskamenicky commented on ListApp/ListApp.Api/Utils/ServerInit.cs:
```diff
@@ -0,0 +1,45 @@
+                }
+            };
+
+            //   config.BindParameter(typeof(Guid), new GuidModelBinder());
```
Remove comments...

tobiaskamenicky commented on listapp/listapp.api.tests/itemscontroller.cs:
```diff
@@ -0,0 +1,349 @@
+﻿using System;
```
why is there a second itemscontroller? Try to remove it from GitHub...

tobiaskamenicky commented on ListApp/ListApp.Api.Tests/ItemsController.cs:
```diff
@@ -0,0 +1,349 @@
+            using (var server = new HttpServer(config))
+            using (var client = new HttpClient(server))
+            {
+                var receivedResponse = await client.GetAsync("http://localhost:57187/api/v1/items/1");
```
I don't think you should test the server...it doesn't really look like a unit test.

tobiaskamenicky commented on ListApp/ListApp.Api.Tests/ItemsController.cs:
```diff
@@ -0,0 +1,349 @@
+        [Test]
+        public async Task Get_ReturnsSpecificExistingItem()
+        {
+            var expectedItem = Utils.Constants.MockListItems.ElementAt(1);
```
I prefer this: 
`  var action = await _theController.GetItem(expectedItem.Id);`
`   var response = await action.ExecuteAsync(CancellationToken.None);`
`   ListItem actual;`
`   response.TryGetContentValue(out actual);`
   
`   Assert.That(response.IsSuccessStatusCode);`

tobiaskamenicky reviewed (changes requested):
I will start with a question: "Have you ever heard of KISS principle?". The task was to create a simple dummy controller, not some filters, patch/batch methods, etc...
Well, you did a great job doing so, but do not complicate things. I mean it doesn't matter in onboarding tasks, at least you practice a little bit, but these things take a lot of time (you are about 3 weeks on this task?) and you will not even need half of the things for the simple TODO app.

Some general comments: 
1. Use R#. There are unused directives that shouldn't be there and some minor things that it can help you to fix.
2. Some comments below are applicable to other places in your code so try to fix them all.
3. Some comments only state that you should change that in the next task, so you don't have to fix that.

Commit:
KA-183 Attempt to remove that weird extra file

Commit:
KA-183 Fix conflicting references

Commit:
KA-183 Rename the controller in tests

Commit:
KA-183 Rename test functions to clarify their behavior

Commit:
KA-183 Move the items initialization into normal constructor

Commit:
KA-183 Add "Async" to the names of controller methods

Commit:
KA-183 Changed GetItemsAsync to return IHttpActionResult

Tests had to be changed accordingly

Commit:
KA-183 Remove extra using directives

Commit:
KA-183 Use WebApiConfig in tests, remove ServerInit

Commit:
KA-183 Refactor the filters to make testable, add unit tests

The filters require injectable HttpRequestMessage, as it is used to generate the error response in case the filter fails. An abstract class with injectable request message was created to base the filters on; each filter now overrides the DoValidation abstract method to its own needs and gets the request object passed into it.

Commit:
KA-183 Remove tests that used server/client to test filters

Until recently, the invalid requests that were supposed to get caught by the action filters were tested by creating instances of HTTP server and client and the corresponding requests were then submitted to the server via the client. In order to stick to unit testing, they have been removed, since the functionality of the action filters is now tested by their own set of unit tests.

Commit:
KA-183 Remove excess namespace specifications

Commit:
KA-183 Refactor test constants names and usage

Commit:
KA-183 Rename controller DeleteItemsCollection method

Commit:
KA-183 Cleanup, rename files, add TODOs
